

Let's list out the GO terms unique to each protein
```{r}
library(dplyr)
library(stringr)
#library


# Takes the Uniprot-derived GO terms for UPS proteins
ups_go <- read.csv('data/uniprot-UPS-go-terms.tab', sep = "\t",
                stringsAsFactors = FALSE) %>%
  select(protein = Entry,
         go_terms = Gene.ontology.IDs)
# Separates them into individual GO terms
x <- 1
ups_go_ind = data.frame(go_terms = character(3000),
                        protein = character(3000),
                        stringsAsFactors = FALSE)
for (n in 1:nrow(ups_go)){
  split_terms <- str_split(
    gsub("\\s","",ups_go$go_terms[n])
      , ";", simplify=TRUE)
  for (m in 1:length(split_terms)){
    ups_go_ind$go_terms[x] <- split_terms[m]
    ups_go_ind$protein[x] <- ups_go$protein[n]
    x = x + 1
  }
}
# Removes rows that have non-unique GO terms
ups_go_ind <- ups_go_ind[!(duplicated(ups_go_ind$go_terms)|duplicated(ups_go_ind$go_terms, fromLast = TRUE)),]
ups_go_ind
```



Let's get the dynamic amounts for each UPS protein
```{r}
ups_ratios <- read.csv("UPS-Protein-Ratios.tsv", sep = "\t",
                header = TRUE, stringsAsFactors = FALSE) %>%
  select(ups2_amount = UPS2.Amount..fmol.,
         protein = UniProt.Accession.Number)
ups_ratios
```


```{r}
invalid <- c("null", "NaN", "Infinity")

# Reads the metaGOmics fold change data
mgo_go <- read.csv('results/metaGOmics/MetaGOmics_Compare_UPS2_UPS1.txt', sep = "\t",
         comment.char="#", header = TRUE, stringsAsFactors = FALSE) %>%
  select(ups1.psms = Run.1.PSM.count,
         ups2.psms = Run.2.PSM.count,
         go_terms = GO.acc,
         fold_change = Log.2..fold.change)

# Join results by GO terms
all_results <- plyr::join_all(list(ups_go_ind, mgo_go), by = "go_terms") %>%
  filter(!is.na(fold_change),
         !fold_change %in% invalid)
all_results <- all_results[rev(order(all_results$fold_change)),]

# print(all_results)

collapsed_results <- all_results
collapsed_results$go_terms <- NULL
collapsed_results <- unique(collapsed_results)
collapsed_results$go_terms <- character(nrow(collapsed_results))

for (i in 1:nrow(collapsed_results)){
  row <- collapsed_results[i,]
  go_terms <- paste(filter(all_results,
                  protein == row$protein,
                  ups1.psms == row$ups1.psms,
                  ups2.psms == row$ups2.psms)$go_terms,
                    collapse = "; ")
  collapsed_results$go_terms[i] <- go_terms
}
#collapsed_results
mgo_go

#proteins <- unique(all_results$protein)
# for (protein in proteins){
#   print(all_results[all_results$protein == protein,])
# }
#ups_go_ind
```




```{r}
# Reads the unipept data
uni1_fp = 'results/unipept/UPS1_03_All_Peptides.csv'
uni2_fp = 'results/unipept/UPS2_03_All_Peptides.csv'

uni1_df <- read.csv(uni1_fp,
         sep = ",", comment.char="", header = TRUE, stringsAsFactors = FALSE) %>%
  select(peptide,
         uni_go_bp = GO..biological.process.,
         uni_go_mf = GO..molecular.function.,
         uni_go_cc = GO..cellular.component.)

uni2_df <- read.csv(uni2_fp,
         sep = ",", comment.char="", header = TRUE, stringsAsFactors = FALSE) %>%
  select(peptide,
         uni_go_bp = GO..biological.process.,
         uni_go_mf = GO..molecular.function.,
         uni_go_cc = GO..cellular.component.)

getCounts <- function(df){
  peptides <- unique(df$peptide)
  n <- length(peptides)
  output <- data.frame(peptide = peptides,
                       count = numeric(n),
                       go_terms = numeric(n))
  for (i in 1:length(peptides)){
      rows <- df[df$peptide == peptides[i],]
      if (length(unique(rows$uni_go_bp)) *
        length(unique(rows$uni_go_cc)) *
        length(unique(rows$uni_go_mf)) == 1){
          output$peptide[i] <- peptides[i]
          output$count[i] <- nrow(rows)
          output$go_terms[i] <- paste(rows[1,]$uni_go_bp,
                                     rows[1,]$uni_go_mf,
                                     rows[1,]$uni_go_cc, collapse = ";")
      }
      else{
        print('Check this!')
      }
    }
  output
}

uni1_df <- getCounts(uni1_df) %>%
  select(peptide,
         ups1_count = count,
         ups1_go_terms = go_terms)
uni2_df <- getCounts(uni2_df) %>%
  select(peptide,
         ups2_count = count,
         ups2_go_terms = go_terms)
```


```{r}
# Runs through the GO terms and sees if any match any UPS GO terms
uni_df <- plyr::join_all(list(uni1_df, uni2_df), by = "peptide") %>%
  filter(ups2_count != 'NA') %>%
  mutate(ratio = ups2_count/ups1_count)
uni_df <- uni_df[rev(order(uni_df$ratio)),]

uni_df$proteins <- numeric(nrow(uni_df))
go_regex <- "(GO:[0-9]+) \\(([0-9]+)%\\)"
for (i in 1:nrow(uni_df)){
  row <- uni_df[i,]
    if (row$ups1_go_terms != row$ups2_go_terms){
      print('Check this: ')
      print(i)
    } else {
      go_terms <- str_match_all(row$ups1_go_terms, go_regex)[[1]]
      proteins <- list()
      if (nrow(go_terms) > 0){
        for (n in 1:nrow(go_terms)){
          #print(go_terms[n,])
          if (as.numeric(go_terms[n, 3]) > 5){ # Keeps terms annotated 5% or above
            if (go_terms[n, 2] %in% ups_go_ind$go_terms){
              protein <- ups_go_ind[ups_go_ind$go_terms == go_terms[n, 2],]$protein
              proteins <- rbind(proteins, protein)
            }
          }
        }
      }
      uni_df$proteins[i] <- paste(unique(proteins), collapse = ";")
      #print(uni_df$proteins)
    }
}
uni_df
# uni_df <- mutate(uni_df,
#                  amounts = ups_ratios[ups_ratios$protein == proteins,]$ups2_amount) %>%
#   filter(uni_df,
#       proteins != "")
```
```{r}
uni_df <- filter(uni_df,
                 proteins %in% ups_ratios$protein)
uni_df
```



```{r}
# MetaProteomeAnalyzer
library("readxl")
mpa_ups1_df <- read_excel("Results/MetaProteomeAnalyzer/UPS1_03_MPA_Analysis.xlsx", sheet = "Proteins") %>%
  select(protein = "Protein Accession",
         ups1_count = "Spectral Count") %>%
  filter(gsub("ups", "", protein) %in% ups_ratios$protein) # TO DO: Change this to look for proteins with the "ups" suffix
  
mpa_ups2_df <- read_excel("Results/MetaProteomeAnalyzer/UPS2_03_MPA_Analysis.xlsx", sheet = "Proteins") %>%
  select(protein = "Protein Accession",
         ups2_count = "Spectral Count") %>%
  filter(gsub("ups", "", protein) %in% ups_ratios$protein) # TO DO: Change this to look for proteins with the "ups" suffix

mpa_ups_df <- plyr::join_all(list(mpa_ups1_df, mpa_ups2_df), by = "protein") %>%
  filter(ups1_count != 'NA',
         ups2_count != 'NA')
mpa_ups_df$protein <- gsub("ups","",mpa_ups_df$protein)

mpa_ups_df <- plyr::join_all(list(mpa_ups_df, ups_ratios), by = "protein") %>%
  mutate(ratio = ups2_count/ups1_count) %>%
  filter(protein %in% ups_ratios$protein)

mpa_ups_df <- mpa_ups_df[rev(order(mpa_ups_df$ratio)),]

mpa_ups_df
#plot(mpa_ups_df$ups2_amount, mpa_ups_df$ups2_count, 
     #xlab="UPS2 Count", ylab= "UPS2 Amount (fmol)")
```

```{r}
#ups_ratios
ups_go_ind[ups_go_ind$protein == "P41159",]
#ups_go_ind[ups_go_ind$protein == "P62988",]
```

